src/lib/backoff.ts

```diff
@@ -0,0 +1,101 @@
+import { strict as assert } from 'assert';
+import { promisify } from 'util';
+
+import { StatusError } from './errors';
+
+// TODO use standard lib async setTimout (requires node 16)
+const sleep = promisify(setTimeout);
+
+export type OnRetryOpts = {
+	retryCount: number;
+	delay: number;
+	error: any;
+};
+
+export type Options = {
+	attempt: number;
+	maxDelay: number;
+	minDelay: number;
+	maxRetries: number;
+	retryCallback?: (info: OnRetryOpts) => void;
+};
+
+const DEFAULT_OPTIONS: Partial<Options> = {
+	attempt: 0,
+	maxDelay: 15000,
+	minDelay: 15000,
+	maxRetries: Infinity,
+};
+
+/**
+ * Retries a function with exponential delay between calls.
+ * If attempts is 0 then, it will delay for min(minDelay, maxDelay)
+ *
+ * Supports exceptions which return `retryAfter` value to specify the backoff duration.
+ * Will call `retryCallback` if passed in options as a way for callers to react to retries
+ */
+export function withBackoff<T extends (...args: any[]) => any>(
+	func: T,
+	options: Partial<Options> = {},
+) {
+	assert(typeof func === 'function', 'expected a function as parameter');
+	// If the function returns a promise, unwrap the promise return type
+	// otherwise use the actual return
+	type TReturn = ReturnType<T> extends Promise<infer R> ? R : ReturnType<T>;
+
+	const normalizedOptions: Options = {
+		...DEFAULT_OPTIONS,
+		...options,
+	} as Options;
+
+	return async function wrapped(...args: Parameters<T>): Promise<TReturn> {
+		try {
+			return await func(...args);
+		} catch (error) {
+			if (isRetryError(error)) {
+				normalizedOptions.attempt = 0;
+			}
+
+			if (++normalizedOptions.attempt >= normalizedOptions.maxRetries) {
```

Comment:
Does `maxRetries = 1`, means you backoff at least once? (i.e. perform 2 tries), or does it mean you just try once? 

If your intention is the first one, I think this should be a `>` instead of a `>=`